src/realm/mixed.cpp

```diff
@@ -85,88 +85,207 @@ inline int compare_float(Float a_raw, Float b_raw)
+inline int compare_generic(T lhs, T rhs)
 {
-    // Comparing types first makes it possible to make a sort of a list of Mixed
-    // This will also handle the case where null values are considered lower than all other values
-    if (m_type > b.m_type)
+    return lhs == rhs ? 0 : lhs < rhs ? -1 : 1;
+}
+
+inline int compare_decimals(Decimal128 lhs, Decimal128 rhs)
+{
+    return lhs.compare(rhs);
+}
+
+inline int compare_decimal_to_double(Decimal128 lhs, double rhs)
+{
+    return lhs.compare(Decimal128(rhs)); // FIXME: slow and not accurate in all cases
+}
+
+// This is the tricky one. Needs to support the following cases:
+// * Doubles with a fractional component.
+// * Longs that can't be precisely represented as a double.
+// * Doubles outside of the range of Longs (including +/- Inf).
+// * NaN (defined by us as less than all Longs)
+// * Return value is always -1, 0, or 1 to ensure it is safe to negate.
+inline int compare_long_to_double(int64_t lhs, double rhs)
+{
+    // All Longs are > NaN
+    if (std::isnan(rhs))
         return 1;
-    else if (m_type < b.m_type)
-        return -1;
 
-    // Now we are sure the two types are the same
+    // Ints with magnitude <= 2**53 can be precisely represented as doubles.
+    // Additionally, doubles outside of this range can't have a fractional component.
+    static const int64_t kEndOfPreciseDoubles = 1ll << 53;
+    if (lhs <= kEndOfPreciseDoubles && lhs >= -kEndOfPreciseDoubles) {
+        return compare_float(double(lhs), rhs);
+    }
+
+    // Large magnitude doubles (including +/- Inf) are strictly > or < all Longs.
+    static const double kBoundOfLongRange = -static_cast<double>(LLONG_MIN); // positive 2**63
+    if (rhs >= kBoundOfLongRange)
+        return -1; // Can't be represented in a Long.
+    if (rhs < -kBoundOfLongRange)
+        return 1; // Can be represented in a Long.
+
+    // Remaining Doubles can have their integer component precisely represented as long longs.
+    // If they have a fractional component, they must be strictly > or < lhs even after
+    // truncation of the fractional component since low-magnitude lhs were handled above.
+    return compare_generic(lhs, int64_t(rhs));
+}
+} // namespace _impl
+
+bool Mixed::is_comparable(const Mixed& l, const Mixed& r)
+{
+    if (l.is_null() || r.is_null())
+        return true;
+
+    DataType l_type = l.get_type();
+    DataType r_type = r.get_type();
+    if (l_type == r_type)
+        return true;
+    bool l_is_numeric = l_type == type_Int || l_type == type_Bool || l_type == type_Float || l_type == type_Double ||
+                        l_type == type_Decimal;
+    bool r_is_numeric = r_type == type_Int || r_type == type_Bool || r_type == type_Float || r_type == type_Double ||
+                        r_type == type_Decimal;
+    if (l_is_numeric && r_is_numeric) {
+        return true;
+    }
+    if ((l_type == type_String && r_type == type_Binary) || (r_type == type_String && l_type == type_Binary)) {
```

Comment:
It's likely expensive to check for (without some prior knowledge), but was it considered to compare strings with numeric in case the strings contains just numbers?

Reply:
Yes, very expensive - and probably not what is wanted.